check_requirements

daniellacosse opened the pull request:
So, using `preinstall` doesn't actually stop things from installing or side effects from happening;

https://user-images.githubusercontent.com/3759828/139708870-21b1ad14-70e0-4aef-8745-787e2af05bee.mov

So i went ahead and made a setup script. Let me know what you think.

daniellacosse commented on scripts/reqcheck/javascript.sh:
```diff
@@ -0,0 +1,39 @@
+source ./scripts/reqcheck/library.sh
+
+# check node version
+LOCAL_NODE_VERSION="$(node --version)"
```
still need to remove the `v` from this

daniellacosse commented on scripts/reqcheck/javascript.sh:
```diff
@@ -0,0 +1,39 @@
+
+# check node version
+LOCAL_NODE_VERSION="$(node --version)"
+NODE_TARGET_VERSION="16.12.0"
```
need to pull this and the npm version below from the `package.json`, along w/ the comparator

daniellacosse replied:
i was thinking, thoughts on a `requirements.ini` or `.requirments.env` file?

daniellacosse commented:
@fortuna once we settle on the approach here i'll merge it and the action proposal PR into the original `daniellacosse/issue-1003` branch, then update the client side branch to match.

fortuna commented on scripts/reqcheck/main.sh:
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash -eu
```
We're overthinking this. The extra infrastructure makes it significantly harder to understand what's going on.

This should do the job:
```sh
#!/bin/bash -eu

function err() {
  echo "$@" >&2
  return 1
}

function main() {
  declare -ri node_version=$(expr "$(node -v)" : 'v\([^.]*\)')
  (( node_version >= 12 )) || err "Need node version >= 12. Found ${node_version}"

  declare -ri npm_version=$(expr "$(npm -v)" : '\([^.]*\)')
  (( npm_version >= 7 )) || err "Need npm version >= 7. Found ${npm_version}"
}

main
```

Let's call the script `check_requirements.sh`. No need to abbreviate. It's not something we will be running often.

I'd like the user to call check_requirements directly, not via npm run, in case npm is not present.

Let's drop the setup script. There's no much point in setting up only if nvm is available, and it's not installing nvm itself.

My suggestion is:
- Use the simple script above.
- Add a line to the requirements section of the readme saying you can run `./scripts/check_requirements.sh` to check the requirements.
- Add the script call to preinstall

daniellacosse replied:
I have tried to build the script with what you said in mind, such that we can extend it down the line more easily, adding different categories of things to check for like we discussed (android, etc). Right now you can check javascript dependencies by passing the `-j` flag, but other flags could be provided later for other groups of dependencies. Just want to confirm you no longer want that functionality.

My issue with the script you have generously provided is that it will diverge from the package.json as the source of truth as we update our node version going forward, as you've hard coded the values. This will cause confusion down the line. It also doesn't respect minor versions (less important).

Really, all I care about is having a core source of truth that we can respect. It doesn't have to be the `package.json` - I am happy to make a `requirements.txt` or something that's easier to understand, and remove node and npm from the `engines` in the package.json so it's clearer.

Perhaps I did it incorrectly, but the intent with calling to nvm in setup was to just switch to the correct version of node if nvm was present for the dev's convenience. `nvm use` looks for an `.nvmrc` and uses that version. It does nothing if nvm is not present, I was just trying to be helpful.

I called it `reqcheck` so it remains consistent with our other checks, e.g. `shellcheck` - is it okay if I rename `shellcheck` to `check_shell`?

Lastly, I can add it to preinstall, but if you watch the video I recorded above, you'll see that `preinstall` is only run as the results of `npm install` are finalized. `node_modules` are still added, and files still written to; I feel `preinstall` gives us a false sense of security and doesn't give us much.

fortuna replied:
Let's establish there's no right or wrong here. :-) We are talking about a tradeoff.

This is adding a lot of code to understand, break and maintain, which is a concern of mine. The amount of code is not proportional to the importance of the problem either. Furthermore, I'd claim that my suggestion is a lot easier to extend, since you can just add a new line or two in the simpler script, without having to understand anything else, and it works with or without a package.json.

It's useful to leave the door open to extensibility in the future, but we should focus on the use case we have. Currently we are just trying to fix npm engine check for new developers that don't read the readme. (But if they don't read the readme they won't even know about `npm run setup`).

I just want a quick fix that will let us move forward with the npm migration (a much more consequential change) and won't add significant overhead.

I hear you about having a single source of truth. In practice I'm ok with having that duplicated in multiple places for now because we rarely change it. Notice that we already duplicate it anyway in the CI config and you're duplicating it in the nvmrc. *If* that turns out to be a real problem, then we can address it. As you make progress with the feature work, you will see we have much bigger fish to fry.

I'm actually comfortable with the strict engines check and the recommendation to use `npm install` (rather than `npm ci`). That should enforce the restriction, right? It may not work with older npm versions, but that's fine. People can figure it out.

fortuna commented on README.md:
```diff
@@ -33,12 +33,12 @@ The system comprises the following components:
 
 In order to build and run the code, you need the following installed:
-  - [Node](https://nodejs.org/)
+  - [Node](https://nodejs.org/) (lts/gallium - `nvm use lts/gallium`)
```
```suggestion
  - [Node 16](https://nodejs.org/en/download/) (lts/gallium)
```

fortuna replied:
Let's specify the number and drop nvm. I don't want people to think that's a requirement.

fortuna commented on README.md:
```diff
@@ -33,12 +33,12 @@ The system comprises the following components:
   - [Wine](https://www.winehq.org/download), if you would like to generate binaries for Windows.
 
-Then you need to install all the NPM package dependencies:
+Then you need to set up the repository with:
```
Revert to npm install

fortuna commented on package.json:
```diff
@@ -13,14 +13,16 @@
     "do": "bash ./scripts/do_action.sh",
-    "actions": "find . -name '*_action.sh' | sed -E 's:./src/(.*)_action.sh:\\1:' | grep -v 'scripts/do_action'",
     "lint": "npm run shellcheck && npm run tslint",
+    "reqcheck": "bash ./scripts/reqcheck/main.sh",
```
Revert these

daniellacosse commented:
Partially addressed the requested changes: 

- removed the setup script in favor of calling `check_requirements.sh` directly. called out `nvm use` in the readme as an alternative.
- renamed `shellcheck` to `check_shell` so it's consistent
- added the requirements check to the pre-commit hook

The rest pend the questions left [here](https://github.com/Jigsaw-Code/outline-server/pull/1008#discussion_r741105822)

fortuna commented on package.json:
```diff
@@ -13,23 +13,23 @@
-    "actions": "find . -name '*_action.sh' | sed -E 's:./src/(.*)_action.sh:\\1:' | grep -v 'scripts/do_action'",
     "lint": "npm run shellcheck && npm run tslint",
-    "shellcheck": "bash ./scripts/shellcheck.sh",
+    "shellcheck": "bash ./scripts/check_shell.sh",
```
The reason why it's called `shellcheck` is because it's the name of the tool. If anything, that's doing linting.

Perhaps we should rename it to `lint:sh`, and `tslint` to `lint:ts`. That's clearer and shows the relationship to `lint`.

daniellacosse replied:
yes! I almost did `check:shell` but `lint:shell` is great too 🥰

daniellacosse commented:
@fortuna since this is now a pretty small change, I'm going to merge it into `daniellacosse/issue-1003` so we can drive this thing home
